src/proto/smb.rs

```diff
@@ -0,0 +1,1143 @@
+
+use chrono;
+
+use crate::client::ClientInfo;
+use crate::Masscanned;
+
+// NBTSession + SMB Header
+// netbios type (1 byte) + reserved (1 byte) + length (2 bytes) + SMB MAGIC (4 bytes)
+//
+pub const SMB1_PATTERN_MAGIC: &[u8; 8] = b"\x00\x00**\xffSMB";
+pub const SMB2_PATTERN_MAGIC: &[u8; 8] = b"\x00\x00**\xfeSMB";
+
+// Build/Dissect secblob with Scapy using: GSSAPI_BLOB(b"`\x82.....")
+const SECURITY_BLOB: &[u8; 320] = b"`\x82\x01<\x06\x06+\x06\x01\x05\x05\x02\xa0\x82\x0100\x82\x01,\xa0\x1a0\x18\x06\n+\x06\x01\x04\x01\x827\x02\x02\x1e\x06\n+\x06\x01\x04\x01\x827\x02\x02\n\xa2\x82\x01\x0c\x04\x82\x01\x08NEGOEXTS\x01\x00\x00\x00\x00\x00\x00\x00`\x00\x00\x00p\x00\x00\x001<*:\xc7+<\xa9m\xac8t\xa7\xdd\x1d[\xf4Rk\x17\x03\x8aK\x91\xc2\t}\x9a\x8f\xe6,\x96\\Q$/\x90MG\xc7\xad\x8f\x87k\"\x02\xbf\xc6\x00\x00\x00\x00\x00\x00\x00\x00`\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\\3S\r\xea\xf9\rM\xb2\xecJ\xe3xn\xc3\x08NEGOEXTS\x03\x00\x00\x00\x01\x00\x00\x00@\x00\x00\x00\x98\x00\x00\x001<*:\xc7+<\xa9m\xac8t\xa7\xdd\x1d[\\3S\r\xea\xf9\rM\xb2\xecJ\xe3xn\xc3\x08@\x00\x00\x00X\x00\x00\x000V\xa0T0R0'\x80%0#1!0\x1f\x06\x03U\x04\x03\x13\x18Token Signing Public Key0'\x80%0#1!0\x1f\x06\x03U\x04\x03\x13\x18Token Signing Public Key";
+
+////////////
+// Common //
+////////////
+
+/// ### PacketDissector
+/// A util class used to dissecate fields.
+#[derive(Debug, Clone)]
+struct PacketDissector<T> {
+    i: usize,
+    state: T,
+}
+impl<T> PacketDissector<T> {
+    fn new(initial_state: T) -> PacketDissector<T> {
+        return PacketDissector {
+            i: 0,
+            state: initial_state,
+        };
+    }
+    fn next_state(&mut self, state: T) {
+        self.state = state;
+        self.i = 0;
+    }
+    fn next_state_when_i_reaches(&mut self, state: T, i: usize) {
+        if self.i == i {
+            self.next_state(state);
+        }
+    }
+    fn _read_u(&mut self, byte: &u8, value: usize, next_state: T, size: usize) -> usize {
+        self.i += 1;
+        self.next_state_when_i_reaches(next_state, size);
+        (value << 8) + *byte as usize
+    }
+    fn _read_ule(&mut self, byte: &u8, value: usize, next_state: T, size: usize) -> usize {
+        let ret = value + ((*byte as usize) << (8 * self.i));
+        self.i += 1;
+        self.next_state_when_i_reaches(next_state, size);
+        ret
+    }
+    fn read_u16(&mut self, byte: &u8, value: u16, next_state: T) -> u16 {
+        self._read_u(byte, value as usize, next_state, 2) as u16
+    }
+    fn read_ule16(&mut self, byte: &u8, value: u16, next_state: T) -> u16 {
+        self._read_ule(byte, value as usize, next_state, 2) as u16
+    }
+    fn read_ule32(&mut self, byte: &u8, value: u32, next_state: T) -> u32 {
+        self._read_ule(byte, value as usize, next_state, 4) as u32
+    }
+    fn read_ule64(&mut self, byte: &u8, value: u64, next_state: T) -> u64 {
+        self._read_ule(byte, value as usize, next_state, 8) as u64
+    }
+}
+
+pub trait Packet {
```

Comment:
This name seems to be unfortunate, as there already is a Packet trait in the library `pnet` that this project uses lots (https://docs.rs/pnet/latest/pnet/packet/trait.Packet.html).

Reply:
Do you think that's an issue? Rust handles quite well naming conflicts.

Reply:
Yes, but if this trait is to be used in other protocol dissectors, I think it could be very confusing for humans, don't you think?

Reply:
Fair enough, I've renamed it MPacket for "Masscanned-packet"